compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNotImplementedOverrideChecker.kt

```diff
@@ -0,0 +1,287 @@
+        val abstractFakeOverrides = potentialFakeOverrides.filter {
+            it is FirMemberDeclaration &&
+                    it.isAbstract &&
+                    // TODO: FIR MPP support
+                    (it.getContainingClass(context) as? FirRegularClass)?.isExpect == false
+        }
+        if (abstractFakeOverrides.isEmpty()) return
+
+        // TODO: differentiate type-substituted declarations. Otherwise, they will be reported as MANY_IMPL_MEMBER_NOT_IMPLEMENTED
+        val sigToDeclarations = mutableMapOf<String, MutableList<FirCallableDeclaration<*>>>()
+        for (abstractFakeOverride in abstractFakeOverrides) {
+            val sig = when (abstractFakeOverride) {
+                is FirSimpleFunction -> SignaturePresenter.represent(abstractFakeOverride)
+                is FirProperty -> SignaturePresenter.represent(abstractFakeOverride)
+                else -> continue
+            }
+            sigToDeclarations.computeIfAbsent(sig) { mutableListOf() }.add(abstractFakeOverride)
+        }
+
+        val canHaveAbstractMembers = declaration is FirRegularClass && declaration.canHaveAbstractFakeOverride
+        // var alreadyReportedManyNotImplemented = false
+        var alreadyReportedAbstractNotImplemented = false
+        for (fakeOverrides in sigToDeclarations.values) {
+            if (fakeOverrides.size > 1) {
+                // TODO: MANY_* as well as some conflict diagnostics
+/*
+                if (alreadyReportedManyNotImplemented) continue
+                val representative = fakeOverrides.first()
+                if (fakeOverrides.any { it.isFromInterface(context) }) {
+                    reporter.reportOn(source, FirErrors.MANY_INTERFACES_MEMBER_NOT_IMPLEMENTED, declaration, representative, context)
+                } else {
+                    reporter.reportOn(source, FirErrors.MANY_IMPL_MEMBER_NOT_IMPLEMENTED, declaration, representative, context)
+                }
+                alreadyReportedManyNotImplemented = true
+*/
+            } else {
+                assert(fakeOverrides.size == 1)
+                val notImplemented = fakeOverrides.single()
+                if ((notImplemented as? FirMemberDeclaration)?.isAbstract != true) continue
+                if (canHaveAbstractMembers || alreadyReportedAbstractNotImplemented) continue
+
+                val notImplementedFunction = notImplemented as? FirSimpleFunction
+                // TODO: suspend function overridden by a Java class in the middle is not properly regarded as an override.
+                if (notImplementedFunction?.isSuspend == true) continue
+                // TODO: Collection extensions are not properly regarded as an override.
+                val originalPackage =
+                    notImplementedFunction?.originalIfFakeOverride()?.dispatchReceiverType?.classId?.packageFqName?.asString()
+                if (originalPackage in COLLECTIONS) continue
```

Comment:
Why is this branching needed? Looks like directly subclassing a Kotlin collection interface would cause execution to return at line 57. Indirect subclassing seems to report fake overrides correctly and this branching is suppressing the errors in such cases.

Reply:
> Why is this branching needed?

As mentioned below, override relations for Collections are not perfectly correct yet at the moment, so to avoid any false positives, I'm bailing out early for now.

> Looks like directly subclassing a Kotlin collection interface would cause execution to return at line 57.

I don't think so. There are `abstract` members in collections, e.g., https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-collection/size.html

So, even direct subclass should implement/override some abstract members. And, if not, fake overrides will be added, and we will catch that here.

> Indirect subclassing seems to report fake overrides correctly and this branching is suppressing the errors in such cases.

Based on how you describe fake overrides for direct/indirect subclasses, I wonder there is a misunderstanding. Without regard to direct/indirect subclass, fake overrides are added if a given class doesn't contribute (i.e., user writes something explicitly) some members that are available through super types. It's simply something that is available at that specific scope (of a class), but not via user's direct contribution.

Anyhow, you're somewhat correct here that we will suppress errors for some cases as well. But, again, it's better not to raise false alarms, which blocks green builds.

Reply:
I'd say that it's important to understand why current check don't work for collections.

Reply:
There weren't that many cases IIRC, so I removed that, and got 2 more correct reports and 3 suspicious cases. Two (out of 3) are kinda expected (?), as described at https://github.com/JetBrains/kotlin/commit/45018ea4685e73040f25d1237b153f882fe22cb3
> incorrect Corner cases like having both contains(Object) and contains(String) within implementation of Collection<String> is not supported

I added TODOs to those test files to not forget. The last one is actually a real issue I think. `Collection#removeAt`, which is a built-in rename (to `remove(index)` I think?) is still detected as a fake override with `abstract` base symbol. Surprised to see that having that fake override in IR doesn't matter (I saw a couple BB tests that will be impacted with that false alarm). Anyway, since it bothers BB tests, I added another kind of ugly whitelist. Not sure we need a generalization for built-in renames or I'm missing something.